### PR TITLE
Implementation of ACE laws, sampling nu, and SIGMA1 doppler broadening. All codes are inspired by OpenMC.

### PR DESCRIPTION
**Pull request by Wenkai-Fu**
1, In NeutronTable class, a function called sample_nu(self, e_in) is added to sample outgoing nu.
        Input: e_in, user-specified incident neutron energy
        Outcome: [nu_total, nu_prompt, nu_delay], if no data is contained in this neutron table, return None, e.g., if only total nu data is presented, return [nu_total, None, None]
1. In Reaction class, a function called sample(self, e_in) is added to sample secondary neutron angle and energy distribution.
        Input: e_in, user-specified incident neutron energy
        Output: [mu, E_out], secondary neutron's angle and energy. In sampling E_out, the applied ace laws are: 3(Inelastic level scattering), 4(Continuous Tabular Distribution), 44(Kalbach-87 Formalism), 61(correlated_sample). Other ace laws can be implemented in the future.
2. Optimize the _get_energy_distributions function in NeutronTable class, includes:
   1. add edist.nd in law 4, 44, 61, so in sample process, code can raise AssertionError if edist.nd != 0, i.e., discrete lines presented.
      1. In law 61, take angle of secondary neutron is isotropic into account, i.e., LC = 0.

### CODE REVIEW

File: pyne/ace.pyx [hunk 1]
```diff
@@ -25,6 +25,7 @@ from collections import OrderedDict
 
 cimport numpy as np
 import numpy as np
+rand = np.random.rand
```
Comment (scopatz) [inline pyne/ace.pyx]: It is more normal to do 

``` python
from numpy.random import rand
```

Comment (Wenkai-Fu) [inline pyne/ace.pyx]: changed

File: pyne/ace.pyx [hunk 2]
```diff
@@ -110,6 +111,79 @@ def ascii_to_binary(ascii_file, binary_file):
     # Close binary file
     binary.close()
 
+def _interpolation_tab1(xin, x, y, interp_NBT = None, interp_INT = None):
+    """
+     INTERPOLATE_TAB1 interpolates a function between two points based on
+     particular interpolation scheme. The data needs to be organized as a ENDF TAB1
+     type function containing the interpolation regions, break points, and
+     tabulated x's and y's.
```
Comment (scopatz) [inline pyne/ace.pyx]: Docstring content should align with quotes

File: pyne/ace.pyx [hunk 3]
```diff
@@ -110,6 +111,79 @@ def ascii_to_binary(ascii_file, binary_file):
...
+    cdef int i, interp  
+    if xin <= x[0]:
+        return y[0]
+    elif xin >= x[-1]:
+        return y[-1]
+    else:
+        I = np.searchsorted(x, xin) - 1
+    
+    # Determine interpolation scheme 
+    if interp_NBT == None:
```
Comment (scopatz) [inline pyne/ace.pyx]: You should almost never do `== None` in Python.  It is horrendously slow and not really want you want.  Please change this to `is None`. Thanks!

Comment (scopatz) [inline pyne/ace.pyx]: Also variable names should be `lowercase_with_underscores` as given in PEP8.  You may also want to check out our style guide. http://pyne.io/devsguide/style_guide.html

File: pyne/ace.pyx [hunk 4]
```diff
@@ -110,6 +111,79 @@ def ascii_to_binary(ascii_file, binary_file):
...
+        r = np.log(xin / x0) / np.log(x1 / x0)
+        return y0 + r * (y1 - y0)
+    elif interp == 4: # log-linear
+        r = (xin - x0) / (x1 - x0)
+        return y0 * np.exp(r * np.log(y1 / y0))
+    elif interp == 5: # log-log
+        r = np.log(xin / x0) / np.log(x1/x0)
+        return y0 * np.exp(r * np.log(y1 / y0))
+    
+def tabular_sample(inter_flag, out, pdf, cdf):
```
Comment (scopatz) [inline pyne/ace.pyx]: Public functions need a documentation string.

File: pyne/ace.pyx [hunk 5]
```diff
@@ -567,16 +641,16 @@ class NeutronTable(AceTable):
             # Polynomial function form of nu
             if LNU == 1:
                 self.nu_t_type = "polynomial"
-                NC = int(self.xss[KNU+1])
-                coeffs = self.xss[KNU+2 : KNU+2+NC]
+                self.nu_t_NC = int(self.xss[KNU+1])
```
Comment (scopatz) [inline pyne/ace.pyx]: PEP8

File: pyne/ace.pyx [hunk 6]
```diff
@@ -588,16 +662,16 @@ class NeutronTable(AceTable):
             # Polynomial function form of nu
             if LNU == 1:
                 self.nu_p_type = "polynomial"
-                NC = int(self.xss[KNU+1])
-                coeffs = self.xss[KNU+2 : KNU+2+NC]
+                self.nu_p_NC = int(self.xss[KNU+1])
+                self.nu_p_coeffs = self.xss[KNU+2 : KNU+2+self.nu_p_NC]
```
Comment (scopatz) [inline pyne/ace.pyx]: I know that this was wrong in the original line, but there should not be whitespace on the sides of the colon, again via PEP8. If you could clean this up, that'd be great!

File: pyne/ace.pyx [hunk 7]
```diff
@@ -659,12 +733,61 @@ class NeutronTable(AceTable):
...
+    def sample_nu(self, e):
+        """
+        Sample nu based on incident neutron energy e, 
+        return a list of length-3, format of [nu_total, nu_prompt, nu_delay].
+        If no nu data is presented for the corresponding nu, return None instead
+        """
+        cdef int i 
+        # No nu data
+        if self.jxs[2] == 0:
+            return [None, None, None]
```
Comment (scopatz) [inline pyne/ace.pyx]: Does this need to be a list, or would a tuple suffice?

File: pyne/ace.pyx [hunk 8]
```diff
@@ -683,8 +805,9 @@ class NeutronTable(AceTable):
                 # No angular distribution data are given for this
                 # reaction, isotropic scattering is asssumed (in CM if
                 # TY < 0 and in LAB if TY > 0)
+                reaction.aflag = 'iso'
                 continue
-
+
```
Comment (scopatz) [inline pyne/ace.pyx]: no need for this line.

File: pyne/ace.pyx [hunk 9]
```diff
@@ -659,12 +733,61 @@ class NeutronTable(AceTable):
...
+        
+        
+        
+        
     def _read_angular_distributions(self):
         """Find the angular distribution for each reaction MT
         """
         cdef int ind, i, j, n_reactions, n_energies, n_bins
-        cdef dict ang_cos, ang_pdf, ang_cdf
+        cdef dict ang_cos, ang_pdf, ang_cdf, JJ
```
Comment (scopatz) [inline pyne/ace.pyx]: PEP8 for `JJ`

File: pyne/ace.pyx [hunk 10]
```diff
@@ -1070,41 +1197,43 @@ class NeutronTable(AceTable):
...
             edist.cdf = []         # Cumulative dist for " " "
+            edist.LC = []          # Use LC to determine ang. data exist or not
 
             npas = []
             edist.a_dist_intt = []
             edist.a_dist_mu_out = [] # Cosine scattering angular grid
             edist.a_dist_pdf = []    # Probability dist function
             edist.a_dist_cdf = []
-            for i in range(NE):
+            for i in range(NE): # for a specific incoming energy i
```
Comment (scopatz) [inline pyne/ace.pyx]: PEP8 two spaces before a comment.

File: pyne/ace.pyx [hunk 11]
```diff
@@ -1573,8 +1711,324 @@ class Reaction(object):
...
+            A = self.table.awr 
+            Q = self.Q
+            E_out = (A / (A + 1.0)) ** 2 * (e - (A + 1.0) / A * Q)
+            mu = self.sample_mu(e)
+            return[mu, E_out]
+            
+        elif edist.law == 4:
+            #Continuous Tabular Distribution 
+            if hasattr(edist, 'INT'):
+                assert len(edist.INT) == 1, 'Multiple interpolation regions not yet supported ' \
```
Comment (scopatz) [inline pyne/ace.pyx]: Raise a real error in the normal way please rather than the assert.

Comment (scopatz) [inline pyne/ace.pyx]: This is a runtime problem, not a data validity problem.

File: pyne/ace.pyx [hunk 12]
```diff
@@ -1573,8 +1711,324 @@ class Reaction(object):
...
+                else:
+                    l = i 
+            else:
+                if (f > rand()):
+                    l = i + 1 
+                else:
+                    l = i 
+                    
+            # check for discrete lines present
+            assert edist.nd[l] == 0, 'Discrete lines in continuous tabular ' \
```
Comment (scopatz) [inline pyne/ace.pyx]: Same.

File: pyne/ace.pyx [hunk 13]
```diff
@@ -1573,8 +1711,324 @@ class Reaction(object):
...
+                if (l == i):
+                    E_out = E_1 + (E_out - E_i_1)*(E_K - E_1)/(E_i_K - E_i_1)
+                else:
+                    E_out = E_1 + (E_out - E_i1_1)*(E_K - E_1)/(E_i1_K - E_i1_1)
+                    
+            # Sample mu 
+            mu = self.sample_mu(e)
+            return [mu, E_out]
+                    
+        elif edist.law == 44:
```
Comment (scopatz) [inline pyne/ace.pyx]: This function is getting pretty unweildy. Please break thi sup into sub-functions.

**Comment by scopatz** [comment]
Thanks a million for putting this in, @Wenkai-Fu! There are many small stylistic changes, but don't be discouraged. Every contributor goes through this on every project at some point. It is normal. 

Once those issues are cleared up, my concern is that this a lack of tests for the new functionality that has been provided.  Now, some of this new functionality is stochastic which may make it harder to test.  However, you can do yourself a favor by setting the random seed in numpy to a static value in the tests. This will make the tests deterministic.  

If you have any questions, please let us know.

**Comment by scopatz** [comment]
Also, CI is failing now even though it was working before. :(

**Comment by paulromano** [comment]
@Wenkai-Fu I can't help but notice that the sampling routines are lifted from OpenMC and Python-ized. That's totally ok but you need to reproduce the copyright/license from OpenMC. @scopatz can advise where the license should go.

**Comment by scopatz** [comment]
Hi @paulromano, with the understanding that I am not a lawyer, this is not my understanding of how copyright works.

The tl;dr is that @Wenkai-Fu ought to have a citation to OpenMC, but should not put in OpenMC's copyright.

In more detail, an algorithm is a sequence of rules. Ideas, rules, facts, and data are uncopyrightable.  What is copyrightable is a concrete expressions of those rules, facts, ideas etc. For example, the rules of Go or Chess are uncopyrightable. However, if I were to publish a book that contained the rules of Chess and Go, the expression (my book, its formatting, typesetting, etc) are copyrightable. To take a programming example, the bubblesort algorithm is not copyrightable. My Python, Java, Haskell, etc implementations of bubblesort are copyrightable.

In this particular case, translation from one programming language to another is a clear re-expression of the idea. As such, it does not fall under the original expression's copyright. No copyright violation is possible because nothing was copied.  Thus I think it would be a mistake to include OpenMC's copyright notice in PyNE for this specific case, because it confers a transfer of copyright to a holder that has no original or transferred claim.

(For the curious, patents hold legal power over ideas, but OpenMC and PyNE are unpatentable at this point.)

All of that being strictly speaking true, OpenMC and PyNE are still works of science and engineering and the provenance of our codes is of utmost importance. Furthermore, failing to appropriately cite work is, again strictly, plagiarism. This may be a relatively minor case, but it is still important.  Personally, I had no idea that this came from OpenMC and I think that put me at a disadvantage.  

I think that the way to resolve this is to put a citation to OpenMC in the docstrings of the added functions to the effect of "this implementation was inspired by the ACE parser in OpenMC vX.X."

Will that work for both of you?

**Comment by Wenkai-Fu** [comment]
These codes were originally a class project, so that is why the format is sloppy and I do not contain the reference statement to OpenMC. But at this point, I should add this info, and sorry about this neglect.  
From my point of view, either reproducing the license or adding a comment in the code is fine for me.

**Comment by scopatz** [comment]
Hi @Wenkai-Fu, you shouldn't reproduce the license for all of the reasons listed above. Please add a comment in the code though.  Let us know when it is ready for another look.

**Comment by Wenkai-Fu** [comment]
Hi @scopatz , could you please help me figure out what is the problem cause the Travis CI failed.  Seems it is a ftp error.

**Comment by paulromano** [comment]
@scopatz Thanks for clarifying. Legalities aside, what I was really getting at is just that there should be proper attribution, which is now the case. Thanks @Wenkai-Fu!

**Comment by scopatz** [comment]
Did the CI issues ever get addressed here?  You may want to merge in master and try it again.

**Comment by Wenkai-Fu** [comment]
Hi @scopatz , the previous CI error resulted from downloading the C12.ace file from ftp folder was auto-resolved. But now the CI is failed with the error message "The command "sudo apt-get install -y gfortran" failed and exited with 100 during ."

**Comment by scopatz** [comment]
Uggh.  Maybe @paulromano or @makeclean or @elliottbiondo or @Pshriwise or someone with Fortran experience can take a look.

**Comment by paulromano** [comment]
Looks like it was probably just a Travis hiccup. I restarted the Travis runs and they seem to be chugging along fine now.

**Comment by makeclean** [comment]
Looks like CI is failing for the same reason as in my PR #826 - I will have a play to see if I can figure out what it is. I think that maybe minconda has its own gfortran version which is older than the gfortran we load at the begining of .travis.yml

**Comment by Wenkai-Fu** [comment]
Hi @scopatz , the implementation of the ACE Laws governing neutron interactions has been written, together with the Doppler broadening of the cross section by SIGMA1 method (inspired by OpenMC). Thanks.

File: pyne/ace.pyx [hunk 14]
```diff
@@ -1025,19 +1195,20 @@ class NeutronTable(AceTable):
...
-                    ND = (INTTp - INTT)/10
-                else:
-                    INTT = INTTp
-                edist.intt.append(INTT)
+                
+                # No matter INTTp > 10 or not, we can get the data 
+                # At sample stage, use nd to determine if there is 
+                # discrete lines error
+                edist.intt.append(INTTp % 10)
+                edist.nd.append(INTTp / 10)
```
Comment (paulromano) [inline pyne/ace.pyx]: This should probably be `INTTp // 10`

File: pyne/ace.pyx [hunk 15]
```diff
@@ -835,6 +1005,7 @@ class NeutronTable(AceTable):
 
             nps = []
             edist.intt = []        # Interpolation scheme (1=hist, 2=lin-lin)
+            edist.nd   = []
```
Comment (paulromano) [inline pyne/ace.pyx]: Maybe add a comment at the end noting this is the number of discrete lines

File: pyne/ace.pyx [hunk 16]
```diff
@@ -1089,22 +1262,21 @@ class NeutronTable(AceTable):
...
-                    INTT = INTTp % 10
-                    ND = (INTTp - INTT)/10
-                else:
-                    INTT = INTTp
-                edist.intt.append(INTT)
-
+                # At sample stage, use nd to determine if there is 
+                # discrete lines present
+                edist.intt.append(INTTp % 10)
+                edist.nd.append(INTTp / 10)
```
Comment (paulromano) [inline pyne/ace.pyx]: Same issue here, `INTTp // 10`

File: pyne/ace.pyx [hunk 17]
```diff
@@ -1555,10 +1735,121 @@ class Reaction(object):
...
+                    b = x[k] - y
+                    self._calculate_f(fb, b)
+                    h = fa - fb
+                    ak = y_inv_sq * h[2] + 2 * y_inv * h[1] + h[0]
+                    bk = y_inv_sq * h[4] + 4 * y_inv * h[3] + 6 * h[2] + \
+                         4 * y * h[1] + y_sq * h[0]
+                    slope = (xs[k] - xs[k - 1]) / (x[k] ** 2 - x[k - 1] ** 2)
+                    sigma = sigma - ak * (xs[k] - slope * x[k] ** 2) - \
+                            slope * bk
+            self.sigmaNew[i] = sigma
```
Comment (paulromano) [inline pyne/ace.pyx]: Personally I think it would be better to just return the broadened cross section rather than add a new attribute

File: pyne/ace.pyx [hunk 18]
```diff
@@ -1555,10 +1735,121 @@ class Reaction(object):
...
+            y = x[i]
+            y_sq = y * y
+            y_inv = 1. / y
+            y_inv_sq = y_inv / y
+            
+            # Evaluate sigma*(y, T) from x[k] - y = 0 to -4
+            k = i
+            a = 0
+            self._calculate_f(fa, a)
+            while (a >= -4 and k > 0):
```
Comment (paulromano) [inline pyne/ace.pyx]: Use floating point on all the 4's, e.g. `a >= -4.0`

File: pyne/ace.pyx [hunk 19]
```diff
@@ -1555,10 +1735,121 @@ class Reaction(object):
...
+                # Move to next point
+                fb = fa.copy()
```
Comment (paulromano) [inline pyne/ace.pyx]: `fb[:] = fa` would avoid create a new array

File: pyne/ace.pyx [hunk 20]
```diff
@@ -1555,10 +1735,121 @@ class Reaction(object):
         self.TY = None     # Neutron release
         self.IE = 0        # Energy grid index
         self.sigma = []    # Cross section values
-
-    def broaden(self, T_high):
-        pass
-
+        self.sqrt_pi_inv = 1. / np.sqrt(np.pi)
```
Comment (paulromano) [inline pyne/ace.pyx]: Don't see why you'd want sqrt_pi_inv to be an instance attribute. I'd just calculate this on the fly or at most, make it a private module variable. Also, I'm pretty sure plain `math.sqrt(math.pi)` is going to be faster than using `np.sqrt` since there is overhead for checking the type of the argument.

File: pyne/ace.pyx [hunk 21]
```diff
@@ -1573,8 +1864,457 @@ class Reaction(object):
...
+                if (abs(mu) > 1):
+                    mu = np.sign(mu)
+            return (mu, E_out)
+        
+    def _maxwell_spectrum(self, t):
+        r1 = rand()
+        r2 = rand()
+        r3 = rand()
+        c = np.cos(.5 * np.pi * r3)
+        e_out = -t * (np.log(r1) + np.log(r2) * c * c)
```
Comment (paulromano) [inline pyne/ace.pyx]: Again, for scalar arguments, `math.log` is probably better

File: pyne/ace.pyx [hunk 22]
```diff
@@ -1573,8 +1864,457 @@ class Reaction(object):
...
+        w = self._maxwell_spectrum(a)
+        e_out = w + .25 * a ** 2 * b + (2 * rand() - 1) * \
+                np.sqrt(a ** 2 * b * w)
+        return e_out
+    
+    def _calculate_f(self, f, a):
+        """Working hourse function used in Doppler broadening
+        """
+        f[0] = .5 * erfc(a)
+        f[1] = .5 * self.sqrt_pi_inv * np.exp(-a * a)
```
Comment (paulromano) [inline pyne/ace.pyx]: Use `math.exp`

File: pyne/ace.pyx [hunk 23]
```diff
@@ -1573,8 +1864,457 @@ class Reaction(object):
...
+        r2 = rand()
+        r3 = rand()
+        c = np.cos(.5 * np.pi * r3)
+        e_out = -t * (np.log(r1) + np.log(r2) * c * c)
+        return e_out
+    
+    def _watt_spectrum(self, a, b):
+        w = self._maxwell_spectrum(a)
+        e_out = w + .25 * a ** 2 * b + (2 * rand() - 1) * \
+                np.sqrt(a ** 2 * b * w)
```
Comment (paulromano) [inline pyne/ace.pyx]: Use `math.sqrt`

File: pyne/ace.pyx [hunk 24]
```diff
@@ -1573,8 +1864,457 @@ class Reaction(object):
...
+        else:
+            E_out = E_1 + (E_out - E_i1_1)*(E_K - E_1)/(E_i1_K - E_i1_1)
+        
+        # Sampled correlated angle from Kalbach-Mann parameters
+        if (rand() > km_r):
+            t = (2.0*rand() - 1.0) * np.sinh(km_a)
+            mu = np.log(t + (t*t + 1.0)**0.5)/km_a
+        else:
+            r1 = rand()
+            mu = np.log(r1*np.exp(km_a) + (1.0 - r1)*np.exp(-km_a))/km_a
```
Comment (paulromano) [inline pyne/ace.pyx]: Use function from `math`

File: pyne/ace.pyx [hunk 25]
```diff
@@ -1573,8 +1864,457 @@ class Reaction(object):
...
+        # Now interpolate between incident energy bins i and i + 1
+        if (l == i):
+            E_out = E_1 + (E_out - E_i_1)*(E_K - E_1)/(E_i_K - E_i_1)
+        else:
+            E_out = E_1 + (E_out - E_i1_1)*(E_K - E_1)/(E_i1_K - E_i1_1)
+        
+        # Sampled correlated angle from Kalbach-Mann parameters
+        if (rand() > km_r):
+            t = (2.0*rand() - 1.0) * np.sinh(km_a)
+            mu = np.log(t + (t*t + 1.0)**0.5)/km_a
```
Comment (paulromano) [inline pyne/ace.pyx]: USe `math.log`

File: pyne/ace.pyx [hunk 26]
```diff
@@ -1573,8 +1864,457 @@ class Reaction(object):
...
+            
+        # Now interpolate between incident energy bins i and i + 1
+        if (l == i):
+            E_out = E_1 + (E_out - E_i_1)*(E_K - E_1)/(E_i_K - E_i_1)
+        else:
+            E_out = E_1 + (E_out - E_i1_1)*(E_K - E_1)/(E_i1_K - E_i1_1)
+        
+        # Sampled correlated angle from Kalbach-Mann parameters
+        if (rand() > km_r):
+            t = (2.0*rand() - 1.0) * np.sinh(km_a)
```
Comment (paulromano) [inline pyne/ace.pyx]: `math.sinh`

File: pyne/ace.pyx [hunk 27]
```diff
@@ -1573,8 +1864,457 @@ class Reaction(object):
...
+                y = -np.log(r1 * r2 * r3)
+            elif edist.nbodies == 5:
+                r1 = rand()
+                r2 = rand()
+                r3 = rand()
+                r4 = rand()
+                r5 = rand()
+                r6 = rand()
+                y = -np.log(r1 * r2 * r3 * r4) - np.log(r5) * \
+                    np.cos(.5 * np.pi * r6) ** 2
```
Comment (paulromano) [inline pyne/ace.pyx]: Use functions from `math`

File: pyne/ace.pyx [hunk 28]
```diff
@@ -1573,8 +1864,457 @@ class Reaction(object):
...
+            if hasattr(edist, 'NBT'):
+                t = _interpolation_tab1(e, edist.energy_in, edist.T, 
+                                         edist.NBT, edist.INT)
+            else:
+                t = _interpolation_tab1(e, edist.energy_in, edist.T)
+            w = (e - edist.U) / t
+            g = 1 - np.exp(-w)
+            e_out = -np.log((1 - g * rand()) * (1 - g * rand()))
+            while e_out > w:
+                e_out = -np.log((1 - g * rand()) * (1 - g * rand()))
```
Comment (paulromano) [inline pyne/ace.pyx]: Use `math.log` and `math.exp` above

File: pyne/ace.pyx [hunk 29]
```diff
@@ -1555,10 +1735,121 @@ class Reaction(object):
         self.TY = None     # Neutron release
         self.IE = 0        # Energy grid index
         self.sigma = []    # Cross section values
-
-    def broaden(self, T_high):
-        pass
-
+        self.sqrt_pi_inv = 1. / np.sqrt(np.pi)
+
+    def broaden(self, t_high):
```
Comment (paulromano) [inline pyne/ace.pyx:1743]: Not sure how much you care about performance, but this method will run _much_ faster with a few static type declarations via Cython

**Comment by paulromano** [comment]
Would be good to have a new test for the `broaden()` method as well

**Comment by Wenkai-Fu** [comment]
Hi @paulromano , do you have any idea about verifying the broadened cross section, e.g., compare the broadened value with your OpenMC results, because I am not familiar with Fortran? Thanks.

**Comment by paulromano** [comment]
@Wenkai-Fu The doppler module that is in OpenMC has not been thoroughly tested and I know that others have discovered issues with it, so I wouldn't consider comparison with OpenMC to be of much value. It would be better to compare to NJOY or, better yet, [Fudge](https://ndclx4.bnl.gov/gf/project/gnd/) (which is open source).

**Comment by scopatz** [comment]
Thanks for updating this @Wenkai-Fu, I'll let @paulromano keep reviewing this. I am a little overwhelmed at the moment.
